### ISSUE TITLE
exp: optimize 1.5B contacts-v1 AFDB + ESM models

### ISSUE THREAD

**Issue by eric-czech** [Open-Athena/MarinFold#199]
## Goal

Run optimization sweeps for model from https://github.com/Open-Athena/MarinFold/issues/155, following the more recent https://github.com/Open-Athena/MarinFold/issues/196 (no crops).

For TRC, start from #117 checkpoints matched to 2 best configs from 3 LR risk profiles (3e-4, 1e-3, 3e-3) implied by https://github.com/Open-Athena/MarinFold/issues/154#issuecomment-5194394007; train w/ cosine LR for 76.1B tokens (~16 AFDB epochs); duplicate trials over 50/50 (uniform ESM/AFDB) and 94/6 (proportional) mix; duplicate again for no augmentation vs 0-100% linear schedule (24 trials total); orchestrate w/ #191 via Opus 5.

For CW, repeat from scratch with WSD (10/70/20) schedule; save 10 permanent checkpoints; train for 2x steps (~32 AFDB epochs); orchestrate w/ #191 via Sol.

## Results

Results are primarily explained in the these evaluation PRs:

- https://github.com/Open-Athena/MarinFold/pull/204
- https://github.com/Open-Athena/MarinFold/pull/219
- https://github.com/Open-Athena/MarinFold/pull/234

<details><summary><b>TODO</b></summary>

- [x] Address `experiments` not exported from `marin`; cannot use `coral` utils (https://github.com/Open-Athena/MarinFold/issues/243)
- [x] Document CW variant w/ WSD and 2x steps
- [x] Document typical checkpoint sizes (https://github.com/eric-czech/marin-agent-kb/blob/main/docs/checkpoint-size.md)
- [x] Explain cooldown prompt: https://gist.github.com/eric-czech/850da8c8c32a7d789f34bc190a8ab6e5
- [x] Explain throughput: https://gist.github.com/eric-czech/6565f00305a0fc708e8b492be7cb295d
- [x] Run cooldown eval: https://wandb.ai/eric-czech/marin/runs/prot-exp199-cw-cv1-p06-cool-s01 (https://github.com/Open-Athena/MarinFold/pull/234)

</details>

<details><summary><b>Implementation</b> (TRC)</summary>

### Baseline

- Implement sweep at `experiments/exp199_optimize_contacts_v1_afdb_esm/exp199_sweep_trc.py`
- Strict starting point: [#166](https://github.com/Open-Athena/MarinFold/issues/166) ([sweep source](https://github.com/marin-community/marin/blob/eac/plm-exp166/experiments/protein/exp166_sweep.py))
- Runtime/source of truth: Marin installed in the MarinFold `.venv`; no other local Marin checkout
- Self-contained experiment script; same Marin dependency surface as [#166](https://github.com/Open-Athena/MarinFold/issues/166)
- Direct `train_lm` calls with `TokenizedCache` inputs; no `Executor`
- MarinFold code copied only where required

### Model

- Architecture inherited unchanged from [#166](https://github.com/Open-Athena/MarinFold/issues/166):
  - Qwen3, ~1.47B parameters
  - `max_seq_len=8192`
  - `hidden_dim=2048`
  - `intermediate_dim=8192`
  - `num_layers=24`
  - `num_heads=32`
  - `num_kv_heads=8`
  - `Llama3RotaryEmbeddingsConfig`
  - QK norm, RMSNorm, SwiGLU, untied embeddings
- Vocabulary: 2,845 tokens, IDs `0–2844`
- Tokenizer: [`eczech/contacts-v1-tokenizer-5d68a24a899f`](https://huggingface.co/eczech/contacts-v1-tokenizer-5d68a24a899f)
  - Stable copy of [`timodonnell/contacts-v1-tokenizer@5d68a24a899f`](https://huggingface.co/timodonnell/contacts-v1-tokenizer/tree/5d68a24a899fdeb3bffd41bed74e121a7b4fe158)
  - No embedding/output-head resizing
- Initialization: corresponding [#117](https://github.com/Open-Athena/MarinFold/issues/117) model checkpoint for every trial
  - Strict model-only load
  - Fresh optimizer, schedule, shuffle, and step zero
  - No scratch-initialized trials

### Data

- Training components from the mixture in [#137](https://github.com/Open-Athena/MarinFold/issues/137) ([source](https://github.com/Open-Athena/MarinFold/blob/claude/contacts-crops-model-training-f6055d/experiments/exp137_models_contacts_and_crops_v1_1_5b/crops_train_common.py#L69))
- AFDB and ESM only; no crops
- Existing token caches only; no copying or retokenization in the experiment DAG
- Missing regional cache: hard failure
- Cache distribution archive: [`eczech/marinfold-exp199-token-caches`](https://huggingface.co/datasets/eczech/marinfold-exp199-token-caches)
- `data_seed=0`
- Block-Feistel shuffle inherited from [#166](https://github.com/Open-Athena/MarinFold/issues/166)
- Sequence packing enabled; prefix-only attention; no cross-document attention
- Target training budget: `76,126,858,749` tokens
  - One complete epoch over the combined AFDB + ESM cache
- Executable training budget:
  - Global batch: 128
  - Sequence length: 8192
  - Tokens per step: `1,048,576`
  - Steps: `72,600`
  - Effective tokens: `76,126,617,600`
  - Step-rounding difference: `-241,149` tokens (`99.999683%` of target)

| ID | AFDB weight | ESM weight | AFDB epochs | ESM epochs |
|---|---:|---:|---:|---:|
| `m1` | 0.500000 | 0.500000 | 8.1388 | 0.5327 |
| `m2` | 0.0614336845 | 0.9385663155 | 0.999997 | 0.999997 |

- `m1`: equal token allocation across unequal component sizes
- `m2`: weights derived directly from cache token counts; approximately one full epoch over each component

### Token caches

| Component | Examples | Tokens | Size |
|---|---:|---:|---:|
| AFDB train | 4,129,682 | 4,676,753,425 | 6.51 GB |
| ESM train | 66,759,922 | 71,450,105,324 | 103.37 GB |
| **Training total** | **70,889,604** | **76,126,858,749** | **109.89 GB** |
| Validation | 41,954 | 47,821,958 | 66.62 MB |
| **Overall total** | **70,931,558** | **76,174,680,707** | **109.95 GB** |

Region-relative cache paths:

- AFDB: `tokenized/contacts-v1/2026.07.13.1`
- ESM: `protein-structure/MarinFold/exp137_contacts_and_crops_v1_1_5b/tokenized/contacts-v1-esm-atlas-train-568225`
- Validation: `tokenized/contacts-v1-val/2026.07.13.1`
- Prefix resolution through `marin_prefix_for_region(REGION)`; no hard-coded bucket names

### Optimization

- AdamW
- Cosine LR schedule
- 10% warmup
- LR and weight decay inherited per [#166](https://github.com/Open-Athena/MarinFold/issues/166) point
- Global batch fixed at 128
- TPU-family-specific tensor parallelism, gradient accumulation, and per-device parallelism inherited from [#166](https://github.com/Open-Athena/MarinFold/issues/166)

### Sweep

Six [#166](https://github.com/Open-Athena/MarinFold/issues/166) points, preserving the corresponding [#117](https://github.com/Open-Athena/MarinFold/issues/117) source checkpoint:

| Point | LR | WD | Source batch | Training batch |
|---|---:|---:|---:|---:|
| `p01` | 3.1623e-3 | 0.2 | 64 | 128 |
| `p02` | 3.1623e-4 | 1.6 | 64 | 128 |
| `p03` | 3.1623e-3 | 0.1 | 128 | 128 |
| `p04` | 1e-3 | 0.8 | 128 | 128 |
| `p05` | 3.1623e-4 | 1.6 | 128 | 128 |
| `p06` | 1e-3 | 0.2 | 64 | 128 |

- Distinct point/checkpoint identity retained for `p02` and `p05`
- Two mixtures: `m1`, `m2`
- Two augmentation variants:
  - `base`: no augmentation
  - `aug`: [#166](https://github.com/Open-Athena/MarinFold/issues/166) amino-acid sequence-statement permutation; per-example probability `0 → 1` over training
- Validation never augmented
- Trial count: `6 points × 2 mixtures × 2 augmentation variants = 24`
- Sweep subversion from CalVer suffix: `s01`, `s02`, …
- Subversion included in W&B identities and checkpoint names

### Evaluation

- Sole validation dataset: `tokenized/contacts-v1-val`
- Full validation cache; no downsampling
- Evaluations every 2,338,376,712 tokens (roughly 1/2 of an AFDB epoch)
- Primary metric: `eval/tokenized/contacts-v1-val/loss`

### Compute

- `REGION` and `TPU` inputs
- Region-agnostic configuration
- Same supported TPU families and placement logic as [#166](https://github.com/Open-Athena/MarinFold/issues/166)
- Region-local caches and initialization checkpoints only
- No cross-region checkpoint or cache reads
- Missing regional prerequisite: hard failure

### IDs and artifacts

W&B run ID:

```text
prot-exp199-cv1-sNN-{m1|m2}-pNN-{base|aug}-{region}
```

Permanent checkpoint root:

```text
checkpoints/protein/prot-exp199-cv1-sNN-{m1|m2}-pNN-{base|aug}
```

- Temporary checkpoints every 10 minutes
- Final permanent checkpoint only
- No experiment outputs at bucket root
- No token-cache writes
- W&B fixed tags: `protein`, `exp199`, `contacts-v1`
- Variable tags: sweep, mixture, point, augmentation, region, TPU, LR, WD, batch, parameters, steps, and tokens
- `smoke` tag only for smoke runs
- `tokens=76126617600` for production runs
- Parameter tag derived from the configured 2,845-token model

### Preflight validation

- Exactly 24 production trials
- Global batch 128 for every trial
- `72,600` training steps for every trial
- All six region-local source checkpoints present and complete
- Strict 2,845-row checkpoint/model vocabulary match
- AFDB, ESM, and validation caches present in the selected region
- Maximum ESM cache token ID below 2,845
- `m2` weights equal cache-token proportions
- Target budget equal to one combined-cache epoch
- Step-rounded token difference recorded explicitly
- Sequence packing enabled
- Validation limited to the complete `contacts-v1-val` cache
- All output paths below `checkpoints/`; all cache paths containing `tokenized/`
- Any violation: stop and discuss before submission

</details>

<details><summary><b>Notes</b></summary>

---

Differences between my experiments using newer Marin APIs (#166) and MarinFold #137:

<img width="781" height="514" alt="Image" src="https://github.com/user-attachments/assets/9264e594-e804-48e6-9396-086db4ba517e" />

---

</details>
